### ISSUE TITLE
Globs don't work consistently for item removal

### ISSUE THREAD

**Issue by tmat** [Microsoft/msbuild#3791]
### Steps to reproduce

Project file
```xml
<Project>
  <ItemGroup>
    <X Include="A1;A2;B"/>
    <X Remove="A*"/>
  </ItemGroup>

  <Target Name="Build">
    <ItemGroup>
      <Y Include="A1;A2;B"/>
      <Y Remove="A*"/>
    </ItemGroup>
    
    <Error Text="@(X)|@(Y)"/>
  </Target>
</Project>
```

### Expected  behavior

error : B|B

### Actual behavior

error : B|A1;A2;B

### Environment data
`msbuild /version` output:

```
Microsoft (R) Build Engine version 15.9.14-preview+gae560092f5 for .NET Framework
Copyright (C) Microsoft Corporation. All rights reserved.

15.9.14.22190
```

**Comment by rainersigwald** [comment]
Unfortunately, the problem is the opposite: globs in item patterns apply _only to files on disk_. They cannot be applied to arbitrary pattern. So it looks like the introduction of the lazy item evaluator (at evaluation time) is applying incorrectly to the pattern, rather than to discovered files on disk.

I'm not sure we can fix that, people may be depending on it (especially since it's intuitive--see the expected behavior in this bug!).

**Comment by tmat** [comment]
I suggest that only glob patterns in `Include` and `Exclude` apply to file paths while a glob in `Remove` applies to what's currently in the item group.

**Comment by rainersigwald** [comment]
That would be a breaking change for inside-target behavior that I'm not willing to make.

**Comment by tmat** [comment]
How about adding a new attribute `RemoveMatching` that would apply globs to the items?

**Comment by rainersigwald** [comment]
That's possible, but seems expensive (in terms of understandability as well as implementation) for the benefit. You can simulate it today with property functions using `string` methods:

```xml
<Y Remove="@(Y)" Condition="$([System.String]::new('%(Y.Identity)').StartsWith('A'))" />
```

**Comment by tmat** [comment]
I find the workaround less understandable. 

What if I want a more complicated pattern, like "*.cs"? I guess I can call Regex. That will make it even more cluttered.

**Comment by rainersigwald** [comment]
The condition could be simplified. https://github.com/Microsoft/msbuild/issues/3799 would be one way to do it.

**Comment by tmat** [comment]
Another scenario: I want to create an item group that lists patterns. Something like "*.dll;*.exe", but doesn't match the file system. Is there some way of escaping wildcards?

**Comment by rainersigwald** [comment]
Yes, it's possible to escape the wildcard. Doc PR to clarify: MicrosoftDocs/visualstudio-docs#1591

Programmatically:

```xml
<I Include="$([MSBuild]::Escape('*.dll'))"/>
```

Manually:

```xml
<I Include="%2A.exe"/> 
```

Note that `Escape` will also escape the semicolon (since that's meaningful to MSBuild too), so you'll need to apply it around each pattern, rather than the whole string.

```
s:\msbuild>type wildcard.proj
<Project>
 <ItemGroup>
    <I Include = "$([MSBuild]::Escape('*.dll'))"/>
    <I Include = "%2A.exe"/>
</ItemGroup>

 <Target Name="X">
  <Warning Text="I: %(I.Identity)" />
 </Target>
</Project>

s:\msbuild>msbuild wildcard.proj
Microsoft (R) Build Engine version 15.9.14-preview+gae560092f5 for .NET Framework
Copyright (C) Microsoft Corporation. All rights reserved.

s:\msbuild\wildcard.proj(8,3): warning : I: *.dll
s:\msbuild\wildcard.proj(8,3): warning : I: *.exe

Build succeeded.
```
